sql/sql_table.cc

```diff
@@ -8453,15 +8454,23 @@ static bool validate_table_encryption(THD *thd, HA_CREATE_INFO *create_info) {
             create_info->tablespace, &tt)) {
       return true;
     }
-    uses_general_tablespace = (tt != Tablespace_type::SPACE_TYPE_IMPLICIT);
+    uses_general_tablespace = (tt != Tablespace_type::SPACE_TYPE_IMPLICIT &&
+                               tt != Tablespace_type::SPACE_TYPE_SHARED);
+    uses_system_tablespace = tt == Tablespace_type::SPACE_TYPE_SYSTEM;
+    if (uses_system_tablespace) {
+      dd::Encrypt_result result = dd::is_system_tablespace_encrypted(thd);
```

Comment:
IMHO, server code shouldn't be talking about InnoDB specific details.

Reply:
am referring to is_system-tablespace_encrypted()

Reply:
I think it is hard to draw a line what is innodb specific and what is not. validate_table_encryption is already calling dd::is_tablespace_encrypted - which returns information whether table is in general tablespace - is this InnoDB specific or not ? I can make is_tablespace_encrypted return uses_general_tablespace = true when system tablespace is used.

Reply:
The encryption validation is done mysql_create_table_no_lock - before innodb is called - thus the encryption must be fixed in the server.

Reply:
General Tablespace is not InnoDB specific concept. Other engines can have genral tablespace too.

I see your point but I think we can avoid specific details here.
From DD perspective, system tablespace is also a general tablespace. A general tablespace is tablespace that can hold multiple tables. Can't we treat system as general type and remove this 'system' specific checks?

Reply:
This also falls under 'get-this-working-and-do-later' category.. So we can definitely improve this. Please file JIRA issue if u want to handle this later.

Reply:
When tablespace is empty and it is not an implicit tablespace (DD treats this as a special case) - is_tablespace_encrypted returns false - so it already discovers that this is system tablespace and it knows that this tablespace is not encrypted. It assumes thus that InnoDB is used - as some other SE could be encrypting system tablespace. 
The only way to determine on DD level whether system tablespace is encrypted or not is by acquiring innodb_system tablespace. So the name innodb_system will shine through the cracks anyways.
So if we want this to be totally transparent - and for any general/system tablespace get answer encrypted yes or no we should delegate all the logic to handlers - which MySQL fails to do, because it already assumes that system tablespace is not encrypted.
However, this is quite big effort and I do not think we should do it now. Probably upstream will implement ibdata1 encryption sooner than later and we would end up removing this code anyways.

Reply:
Ok, fine to defer this

Reply:
Thanks for explaining the current logic.